flink-runtime/src/test/java/org/apache/flink/runtime/state/InternalPriorityQueueTestBase.java

```diff
@@ -0,0 +1,452 @@
+
+	protected static void insertRandomTimers(
+		@Nonnull InternalPriorityQueue<TestElement> priorityQueue,
+		@Nonnull Set<TestElement> checkSet,
+		int count) {
+
+		ThreadLocalRandom localRandom = ThreadLocalRandom.current();
+
+		final int numUniqueKeys = Math.max(count / 4, 64);
+
+		long duplicatePriority = Long.MIN_VALUE;
+
+		for (int i = 0; i < count; ++i) {
+			TestElement element;
+			do {
+				long elementPriority;
+				if (duplicatePriority == Long.MIN_VALUE) {
+					elementPriority = localRandom.nextLong();
+				} else {
+					elementPriority = duplicatePriority;
+					duplicatePriority = Long.MIN_VALUE;
+				}
+				element = new TestElement(localRandom.nextInt(numUniqueKeys), elementPriority);
+			} while (!checkSet.add(element));
+
+			if (localRandom.nextInt(10) == 0) {
+				duplicatePriority = element.getPriority();
+			}
+
+			final boolean headChangedIndicated = priorityQueue.add(element);
+			if (element.equals(priorityQueue.peek())) {
+				Assert.assertTrue(headChangedIndicated);
+			}
+		}
+		Assert.assertEquals(count, priorityQueue.size());
+	}
+
+	@Test
+	public void testPeekPollOrder() {
+		final int initialCapacity = 4;
+		final int testSize = 1000;
+		InternalPriorityQueue<TestElement> priorityQueue =
+			newPriorityQueue(initialCapacity);
+		HashSet<TestElement> checkSet = new HashSet<>(testSize);
+
+		insertRandomTimers(priorityQueue, checkSet, testSize);
+
+		long lastPriorityValue = Long.MIN_VALUE;
+		int lastSize = priorityQueue.size();
+		Assert.assertEquals(testSize, lastSize);
+		TestElement testElement;
+		while ((testElement = priorityQueue.peek()) != null) {
+			Assert.assertFalse(priorityQueue.isEmpty());
+			Assert.assertEquals(lastSize, priorityQueue.size());
+			Assert.assertEquals(testElement, priorityQueue.poll());
+			Assert.assertTrue(checkSet.remove(testElement));
+			Assert.assertTrue(testElement.getPriority() >= lastPriorityValue);
+			lastPriorityValue = testElement.getPriority();
+			--lastSize;
+		}
+
+		Assert.assertTrue(priorityQueue.isEmpty());
+		Assert.assertEquals(0, priorityQueue.size());
+		Assert.assertEquals(0, checkSet.size());
+	}
+
+	@Test
+	public void testStopInsertMixKeepsOrder() {
+
+		InternalPriorityQueue<TestElement> priorityQueue = newPriorityQueue(3);
+
+		final int testSize = 128;
+		HashSet<TestElement> checkSet = new HashSet<>(testSize);
+
+		insertRandomTimers(priorityQueue, checkSet, testSize);
+
+		// check that the whole set is still in order
+		while (!checkSet.isEmpty()) {
+
+			Iterator<TestElement> iterator = checkSet.iterator();
+			TestElement element = iterator.next();
+			iterator.remove();
+
+			boolean removesHead = element.equals(priorityQueue.peek());
+			if (removesHead) {
+				Assert.assertTrue(priorityQueue.remove(element));
+			} else {
+				priorityQueue.remove(element);
+			}
+			Assert.assertEquals(checkSet.size(), priorityQueue.size());
+
+			long lastPriorityValue = Long.MIN_VALUE;
```

Comment:
Could be
`long lastPriorityValue = removesHead ? element.getPriority() : Long.MIN_VALUE;`

Reply:
👍